core/server/master/src/main/java/alluxio/master/meta/checkconf/ServerConfigurationChecker.java

```diff
@@ -0,0 +1,184 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master.meta.checkconf;
+
+import alluxio.PropertyKey;
+import alluxio.PropertyKey.ConsistencyCheckLevel;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.function.Function;
+
+/**
+ * This class is responsible for checking server-side configuration.
+ */
+public class ServerConfigurationChecker {
+  /** Contain all the master configuration information. */
+  private final ServerConfigurationRecord mMasterRecord;
+  /** Contain all the worker configuration information. */
+  private final ServerConfigurationRecord mWorkerRecord;
+
+  /** Listeners to call when need to get master hostname. */
+  private final BlockingQueue<Function<Long, String>> mGetMasterHostnameListeners
+      = new LinkedBlockingQueue<>();
+  /** Listeners to call when need to get worker hostname. */
+  private final BlockingQueue<Function<Long, String>> mGetWorkerHostnameListeners
+      = new LinkedBlockingQueue<>();
+
+  /** Record the configuration errors of last check conf. */
+  private List<WrongProperty> mConfErrors;
+  /** Record the configuration warnings of last check conf. */
+  private List<WrongProperty> mConfWarns;
+
+  /**
+   * Status of the check.
+   */
+  public enum Status {
+    PASSED, // do not have configuration errors and warnings
+    WARN, // do not have configuration errors but have warnings
+    FAILED, // have configuration errors
+    NOT_STARTED,
+  }
+
+  /** Record the status of last check conf. */
+  private Status mStatus;
+
+  /**
+   * Constructs a new {@link ServerConfigurationChecker}.
+   *
+   * @param masterRecord master configuration record
+   * @param workerRecord worker configuration record
+   */
+  public ServerConfigurationChecker(ServerConfigurationRecord masterRecord,
+      ServerConfigurationRecord workerRecord) {
```

Comment:
how will we synchronize access to these? They could be modified from outside this class while this class is in the middle of reading them.

Reply:
You mean masterRecord and workerRecord may be modified from outside the class?
All the masterConfigRecord, workerConfigRecord and configChecker are final inside the AlluxioMasterProcess class. 

ConfigChecker needs to get configuration from masterConfigRecord and workerConfigRecord.
Each time we call checkConf after updating the records. checker.checkConf will get conf map from records. record.getConfMap itself is synchronized. So it will be executed after updating the records. They are sequential.

Reply:
Got it, getConfMap is synchronized so it's safe to use `ServerConfigurationRecord` within `ServerConfigurationChecker` as long as we only make one call to `getConfMap`.